### DISCUSSION TITLE
What is the syntax for the shipment allocation API?

### DISCUSSION THREAD

**Discussion by miggland** [inventree/InvenTree#3256] [Q&A] [answered]
I'm trying out the shipment allocation API, at:

I'm using the browser REST interface, and sending this:
```
{
    "items": [27],
    "shipment": 7
}
```

This gives me this answer on InvenTree 0.7.4:
```
{
    "items": [
        {
            "non_field_errors": [
                "Invalid data. Expected a dictionary, but got int."
            ]
        }
    ]
}
```

On my test server, I'm getting this reply (on latest tag, updated yesterday, 24th June):
```
{
    "items": [
        "This field is required."
    ],
    "shipment": [
        "This field is required."
    ]
}
```
What is the expected syntax here?

The api-doc page is not very helpful:
![image](https://user-images.githubusercontent.com/1314853/175764866-3e42ea65-bddf-40a2-9ccb-272dc3f34cf6.png)

I'm assuming items is a list of PKs for the so-line-items?

**Comment by matmair** [comment, marked as answer]
Hi there @miggland . As the error says you need to supply a dictionary, not an integer. Here is a sample request:
![grafik](https://user-images.githubusercontent.com/66015116/175791279-c8dde17f-ef0d-431a-9bf7-625e397fdf8a.png)

**Reply by miggland** [reply]
Thanks - I did see that it expected a dictionary, but I had no clue what to put inside it. What is the right place to put the description for this, so that it shows up in /api-docs/ ?

**Reply by matmair** [reply]
@miggland no idea, maybe @SchrodingersGat has input

**Reply by SchrodingersGat** [reply]
It looks like the current schema generation is limited in that it does not document child fields. We are going to update the API schema generation in the next release cycle - https://github.com/inventree/InvenTree/issues/3226 - so will address this issue then.

**Comment by miggland** [comment]
I'm having trouble with this in the inventree-python binding..

Using the api.post method, I get this:
```
data = {'items': [{'line_item': 1, 'quantity': '3', 'stock_item': '231'}], 'shipment': 1}
type(data['items']) = <class 'list'>

self._api.post(url, data)

POST request failed at 'http://ubuntuvm:8000/api/order/so/1/allocate/' - 400
{"items":{"non_field_errors":["Expected a list of items but got type \"str\"."]}}
```

However, if I send the same data with:
`response = requests.post(URL, json=data, headers=headers)`, this error does not appear.

Is anybody clear about when data, and when json should be used? The rest of the API seems to work, as this is not the only place JSON data is sent? Or does it have to do with the nested JSON?

This is using inventree-dev:
```
# Version Information:
InvenTree-Version: 0.8.0 dev
Django Version: 3.2.14
Commit Hash: 6e2a2f2b2
Commit Date: 2022-07-15
Database: sqlite3
Debug-Mode: True
Deployed using Docker: True
```

**Comment by SchrodingersGat** [comment]
Hey @miggland I've had a look into this - let's discuss over at https://github.com/inventree/inventree-python/pull/115
